include/treelite/tree_impl.h

```diff
@@ -705,26 +788,71 @@ Model::Dispatch(Func func) const {
 inline std::vector<PyBufferFrame>
 Model::GetPyBuffer() {
   std::vector<PyBufferFrame> buffer;
+  buffer.push_back(GetPyBufferFromScalar(&major_ver_));
+  buffer.push_back(GetPyBufferFromScalar(&minor_ver_));
+  buffer.push_back(GetPyBufferFromScalar(&patch_ver_));
   buffer.push_back(GetPyBufferFromScalar(&threshold_type_));
   buffer.push_back(GetPyBufferFromScalar(&leaf_output_type_));
   this->GetPyBuffer(&buffer);
   return buffer;
 }
 
 inline std::unique_ptr<Model>
-Model::CreateFromPyBuffer(std::vector<PyBufferFrame> frames) {
+Model::CreateFromPyBufferImpl(std::vector<PyBufferFrame> frames, bool assume_ownership) {
+  int major_ver, minor_ver, patch_ver;
   TypeInfo threshold_type, leaf_output_type;
-  if (frames.size() < 2) {
-    throw std::runtime_error("Insufficient number of frames: there must be at least two");
+  constexpr size_t kNumFrameInHeader = 5;
+  if (frames.size() < kNumFrameInHeader) {
+    throw std::runtime_error(std::string("Insufficient number of frames: there must be at least ")
+                             + std::to_string(kNumFrameInHeader));
+  }
+  InitScalarFromPyBuffer(&major_ver, frames[0], assume_ownership);
+  InitScalarFromPyBuffer(&minor_ver, frames[1], assume_ownership);
+  InitScalarFromPyBuffer(&patch_ver, frames[2], assume_ownership);
+  if (major_ver != TREELITE_VER_MAJOR || minor_ver != TREELITE_VER_MINOR) {
+    throw std::runtime_error("Cannot deserialize model from a different version of Treelite");
   }
-  InitScalarFromPyBuffer(&threshold_type, frames[0]);
-  InitScalarFromPyBuffer(&leaf_output_type, frames[1]);
+  InitScalarFromPyBuffer(&threshold_type, frames[3], assume_ownership);
+  InitScalarFromPyBuffer(&leaf_output_type, frames[4], assume_ownership);
 
   std::unique_ptr<Model> model = Model::Create(threshold_type, leaf_output_type);
-  model->InitFromPyBuffer(frames.begin() + 2, frames.end());
+  model->InitFromPyBuffer(frames.begin() + kNumFrameInHeader, frames.end(), assume_ownership);
   return model;
 }
 
+inline std::unique_ptr<Model>
+Model::CreateFromPyBuffer(std::vector<PyBufferFrame> frames) {
```

Comment:
Is the copy necessary here? Is there a scenario in which someone would call `CreateFromPyBuffer(frames)` and then continue to use those frames later? Could we just accept `std::vector<PyBufferFrame>&&` instead?

Reply:
> Is there a scenario in which someone would call `CreateFromPyBuffer(frames)` and then continue to use those frames later?

Yes, the `frames` can originate from outside. For example, in MNMG setting, each worker process receives a series of frames that represent a cuML RF model. The frames are held by the Python interpreter, and our C++ code has no control over when they will be used or destroyed.